### PR TITLE
mintResource

### PR DESCRIPTION
**Pull request by pixelexe**
implementation: metaTemplate, mintResource, NFCController (write page) Change pro/eleveur/naissance
To be check:
 Write

### CODE REVIEW

File: gui/symfony-docker/templates/user/ScanNFC.html.twig [hunk 1]
```diff
@@ -0,0 +1,17 @@
+{% extends 'template.html.twig' %}
+
+{% block body %}
+
+<div class="container-fluid">
+    <h1>Présenter votre TagNFC près du lecteur </h1><br>
```
Comment (qtambone) [inline gui/symfony-docker/templates/user/ScanNFC.html.twig]: this is the good wording to use : 

"Présenterz un tag NFC à proximité du scanner, puis cliquer sur le bouton "Ecrire" pour associer y associer la ressource créée"

File: gui/symfony-docker/src/Service/HardwareService.php [hunk 2]
```diff
@@ -67,4 +66,85 @@ public function startReader(): JsonResponse {
         return null;
     }
 
+    public function mintResource(int $resourceId,
```
Comment (qtambone) [inline gui/symfony-docker/src/Service/HardwareService.php]: This function has no place in the HardwareService, as it concerns the blockchain, so you need to create a specific service, for example: BlockchainService

File: gui/symfony-docker/src/Service/HardwareService.php [hunk 3]
```diff
@@ -67,4 +66,85 @@ public function startReader(): JsonResponse {
...
+            ]);
+
+            return $response->getContent(); // Obtenir le JSON brut
+        } catch (\Exception $e) {
+            return "Error";
+            // Handle exception
+        }
+    }
+
+    public function metadataTemplate(int $weight = 0,
```
Comment (qtambone) [inline gui/symfony-docker/src/Service/HardwareService.php]: This function has no place in the HardwareService, as it concerns the blockchain metadata when created a resource, so you need to put it in the blockchain specific service, for example: BlockchainService

File: gui/symfony-docker/src/Controller/NFCController.php [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+    private HardwareService $hardwareService;
+
+    public function __construct(HardwareService $hardwareService)
+    {
+        $this->hardwareService = $hardwareService;
+        
+    }
+
+    #[Route('/NFC/write/{id}', name: 'app_nfc_write')]
+    public function ScanNFC(Request $request, $id): Response
```
Comment (qtambone) [inline gui/symfony-docker/src/Controller/NFCController.php]: Why is this function called "scanNFC" if its purpose is to do a write? Call it "startWrite" please.

File: gui/symfony-docker/src/Controller/NFCController.php [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+{
+    if ($request->isMethod('POST')) {       //if post method
+        // The form is submitted
+        $response = $this->hardwareService->write($id);
+        // Do something with the data
+        if ($response->getStatusCode() === 200) {
+            $this->addFlash('success', 'The data has been saved');
```
Comment (qtambone) [inline gui/symfony-docker/src/Controller/NFCController.php]: replace by : ('success', 'The NFT ID of the resource has been write on the NFC')

File: gui/symfony-docker/src/Controller/NFCController.php [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+        } else {
+            $this->addFlash('error', 'An error occurred while saving the data');
```
Comment (qtambone) [inline gui/symfony-docker/src/Controller/NFCController.php]: replace by : ('error', 'An error occurred while writing the resource NFT ID on the NFC')

File: gui/symfony-docker/src/Controller/EleveurController.php [hunk 7]
```diff
@@ -48,27 +63,38 @@ public function index(): Response
...
+            // dd($form->getData());
+            $metadata = $this->hardwareService->metadataTemplate((int)$form->getData()->getWeight(),
+                                                                (int)$form->getData()->getPrice(),
+                                                                $form->getData()->getDescription(),
+                                                                $form->getData()->getGenre()
+                                                            );
+            $response = $this->hardwareService->mintResource(1,1, ['metadata' => $metadata]);
+            $responseArray = json_decode($response, true);
+            // dd($responseArray);
+            // $this->addFlash('success', 'La naissance de votre animal a bien été enregistrée ! NFC : ' . $responseArray[1]['tokenId']);
```
Comment (qtambone) [inline gui/symfony-docker/src/Controller/EleveurController.php]: here it is the NFT that is created on the blockchain (mintResource), so replace the "NFC" in the addFlash with "NFT".

File: gui/symfony-docker/src/Controller/EleveurController.php [hunk 8]
```diff
@@ -48,27 +63,38 @@ public function index(): Response
...
+        } catch (UniqueConstraintViolationException){
+            $this->addFlash('error', 'Le tag NFC est déjà utilisé');
```
Comment (qtambone) [inline gui/symfony-docker/src/Controller/EleveurController.php]: same here, we're talking about NFT, not NFC. 
(please check the difference between an NFT resource and an NFC tag to be clear for the next PR)

File: gui/symfony-docker/src/Controller/EleveurController.php [hunk 9]
```diff
@@ -48,27 +63,38 @@ public function index(): Response
...
+            return $this->redirectToRoute('app_eleveur_naissance');
         }
-        return $this->render('pro/eleveur/naissance.html.twig', [
-            'form' => $form->createView(),
-        ]);
+        // $this->addFlash('success', 'La naissance de votre animal a bien été enregistrée !');
+        $this->addFlash('success', 'La naissance de votre animal a bien été enregistrée ! NFC : ' . $responseArray[1]['tokenId']);
```
Comment (qtambone) [inline gui/symfony-docker/src/Controller/EleveurController.php]: NFC -> to be replace by "NFT"

File: gui/symfony-docker/src/Controller/EleveurController.php [hunk 10]
```diff
@@ -48,27 +63,38 @@ public function index(): Response
...
+    $form->handleRequest($request);
+    if ($form->isSubmitted() && $form->isValid()) {
+        try {
+            // dd($form->getData());
+            $metadata = $this->hardwareService->metadataTemplate((int)$form->getData()->getWeight(),
+                                                                (int)$form->getData()->getPrice(),
+                                                                $form->getData()->getDescription(),
+                                                                $form->getData()->getGenre()
+                                                            );
+            $response = $this->hardwareService->mintResource(1,1, ['metadata' => $metadata]);
```
Comment (qtambone) [inline gui/symfony-docker/src/Controller/EleveurController.php]: so here you will need to use the templates that are saved in the contract. you do not know what ressource is behind the nummero 1. 

what is needed: 
1. getTemplates that return the templates from the contract (implement on the BC24-API side)
2. filter those for role BREEDER
3. provide the user with a dropdown to select the animal he wants to create
4. from this selection you will be able to get the templateID

**Review by qtambone — CHANGES_REQUESTED** [review]
i let you some comments.

File: gui/symfony-docker/src/Controller/EleveurController.php [hunk 11]
```diff
@@ -75,21 +77,18 @@ public function naissance(Request $request,
...
                                                                 $form->getData()->getGenre()
                                                             );
-            $response = $this->hardwareService->mintResource(1,1, ['metadata' => $metadata]);
+            $response = $this->blockChainService->mintResource(1,1, ['metadata' => $metadata]);
             $responseArray = json_decode($response, true);
-            // dd($responseArray);
-            // $this->addFlash('success', 'La naissance de votre animal a bien été enregistrée ! NFC : ' . $responseArray[1]['tokenId']);
         } catch (UniqueConstraintViolationException){
-            $this->addFlash('error', 'Le tag NFC est déjà utilisé');
+            $this->addFlash('error', 'Le tag NFT est déjà utilisé');
```
Comment (qtambone) [inline gui/symfony-docker/src/Controller/EleveurController.php]: Ce message d'erreur n'a rien à voir, à remplacer par :    $this->addFlash('error', 'Le NFT existe déjà');

**Review by qtambone — CHANGES_REQUESTED** [review]

**Review by qtambone — APPROVED** [review]

File: gui/symfony-docker/templates/user/ScanNFC.html.twig
Comment (qtambone) [inline gui/symfony-docker/templates/user/ScanNFC.html.twig:1]: please use a correct name for this file, it is not about "ScanNFC", it is more "WriteOnNFC"

File: gui/symfony-docker/templates/user/ScanNFC.html.twig [hunk 12]
```diff
@@ -3,7 +3,7 @@
 {% block body %}
 
 <div class="container-fluid">
-    <h1>Présenter votre TagNFC près du lecteur </h1><br>
+    <h1>Présenterz un tag NFC à proximité du scanner, puis cliquer sur le bouton "Ecrire" </h1><br>
```
Comment (qtambone) [inline gui/symfony-docker/templates/user/ScanNFC.html.twig]: my bad, a "z" typo is here

**Review by qtambone — CHANGES_REQUESTED** [review]

File: gui/symfony-docker/src/Controller/EleveurController.php [hunk 13]
```diff
@@ -71,26 +71,29 @@ public function index(): Response
...
+                                                                $form->getData()["description"],
+                                                                $form->getData()["Genre"],
+                                                                false
                                                             );
-            $response = $this->blockChainService->mintResource(1,1, ['metadata' => $metadata]);
+            $response = $this->blockChainService->mintResource((int)$form->getData()["resourceName"],1, ['metadata' => $metadata]);
             $responseArray = json_decode($response, true);
         } catch (UniqueConstraintViolationException){
-            $this->addFlash('error', 'Le tag NFT est déjà utilisé');
+            $this->addFlash('error', 'Le tag NFC est déjà utilisé');
```
Comment (qtambone) [inline gui/symfony-docker/src/Controller/EleveurController.php]: we don't talk about NFC here. mint resource is about NFT, so please use good error message.. 
the good one is : ('error', 'Le NFT existe déjà')

**Review by qtambone — CHANGES_REQUESTED** [review]

File: gui/symfony-docker/src/Controller/EleveurController.php (hunk 13 above)
Comment (pixelexe) [reply gui/symfony-docker/src/Controller/EleveurController.php]: done sorry

**Review by qtambone — APPROVED** [review]

**Pull request merged**